sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -614,9 +631,22 @@ pi_result piDevicesGet(pi_platform Platform, pi_device_type DeviceType,
 
   // Get number of devices supporting Level Zero
   uint32_t ZeDeviceCount = 0;
+  std::lock_guard<std::mutex> Lock(Platform->PiDevicesCacheMutex);
+  if (Platform->CacheInvalidated) {
+    for (const pi_device CachedDevice : Platform->PiDevicesCache) {
```

Comment:
This looks weird that you iterate an invalid cache. What exactly is not valid?

Reply:
As soon as there is a call to piDeviceRelease for any pi_device that belongs to pi_platform, we invalidate the entire cache.
Probably, I need to remind you what we discussed before.
You asked me "if we invalidate only one pi_device, we will not return the whole pi_devices since we only return partial pi_devices in the cache."
So, I fixed this issue by invalidating the entire cache in piDeviceRelease().
When there a  new request for piDevicesGet for this pi_platform, we revive the cache by calling initialize().
This is cheaper than physically removing the cached pi_devices and calling again L0 RT to get Ze_Device_Handles afresh.

Reply:
added more comments about how invalidation works between piDeviceRelease and piDevicesGet.